File: src/model/GameModel.ts
import { Timestamp } from "firebase/firestore";

export interface IGameModel {
  id: string;
  title: string;
  description: string;
  levelNumber: number;
  videoUrl: string;
  thumbnailUrl?: string | null;
  estimatedTime: number;
  accessTier: string;
  isActive: boolean;
  sortOrder: number;
  createdAt: Date;
  updatedAt: Date;
  createdBy: string;
  category: string;
  tipText: string;
  videoDurationSeconds: number;
  levelId: string;
}

export class GameModel implements IGameModel {
  id: string;
  title: string;
  description: string;
  levelNumber: number;
  videoUrl: string;
  thumbnailUrl?: string | null;
  estimatedTime: number;
  accessTier: string;
  isActive: boolean;
  sortOrder: number;
  createdAt: Date;
  updatedAt: Date;
  createdBy: string;
  category: string;
  tipText: string;
  videoDurationSeconds: number;
  levelId: string;

  constructor(data: IGameModel) {
    this.id = data.id;
    this.title = data.title;
    this.description = data.description;
    this.levelNumber = data.levelNumber;
    this.videoUrl = data.videoUrl;
    this.thumbnailUrl = data.thumbnailUrl;
    this.estimatedTime = data.estimatedTime;
    this.accessTier = data.accessTier;
    this.isActive = data.isActive;
    this.sortOrder = data.sortOrder;
    this.createdAt = data.createdAt;
    this.updatedAt = data.updatedAt;
    this.createdBy = data.createdBy;
    this.category = data.category;
    this.tipText = data.tipText;
    this.videoDurationSeconds = data.videoDurationSeconds;
    this.levelId = data.levelId;
  }

  /** Factory: create a new GameModel instance */
  static create(params: {
    id: string;
    title: string;
    description: string;
    levelNumber: number;
    videoUrl: string;
    thumbnailUrl?: string | null;
    estimatedTime: number;
    accessTier?: string;
    isActive?: boolean;
    sortOrder?: number;
    createdBy?: string;
    category: string;
    tipText: string;
    videoDurationSeconds: number;
    levelId: string;
  }): GameModel {
    const now = new Date();
    return new GameModel({
      id: params.id,
      title: params.title,
      description: params.description,
      levelNumber: params.levelNumber,
      videoUrl: params.videoUrl,
      thumbnailUrl: params.thumbnailUrl ?? null,
      estimatedTime: params.estimatedTime,
      accessTier: params.accessTier ?? "free",
      isActive: params.isActive ?? true,
      sortOrder: params.sortOrder ?? 0,
      createdAt: now,
      updatedAt: now,
      createdBy: params.createdBy ?? "",
      category: params.category,
      tipText: params.tipText,
      videoDurationSeconds: params.videoDurationSeconds,
      levelId: params.levelId,
    });
  }

  /** Convert to Firestore JSON */
  toFirestore(): Record<string, any> {
    return {
      id: this.id,
      title: this.title,
      description: this.description,
      levelNumber: this.levelNumber,
      videoUrl: this.videoUrl,
      thumbnailUrl: this.thumbnailUrl,
      estimatedTime: this.estimatedTime,
      accessTier: this.accessTier,
      isActive: this.isActive,
      sortOrder: this.sortOrder,
      createdAt: Timestamp.fromDate(this.createdAt),
      updatedAt: Timestamp.fromDate(this.updatedAt),
      createdBy: this.createdBy,
      category: this.category,
      tipText: this.tipText,
      videoDurationSeconds: this.videoDurationSeconds,
      levelId: this.levelId,
    };
  }

  /** Convert Firestore doc → GameModel */
  static fromFirestore(json: Record<string, any>): GameModel {
    return new GameModel({
      id: json.id ?? "",
      title: json.title ?? "",
      description: json.description ?? "",
      levelNumber: json.levelNumber ?? 1,
      videoUrl: json.videoUrl ?? "",
      thumbnailUrl: json.thumbnailUrl ?? null,
      estimatedTime: json.estimatedTime ?? 0,
      accessTier: json.accessTier ?? "free",
      isActive: json.isActive ?? true,
      sortOrder: json.sortOrder ?? 0,
      createdAt: GameModel._toDateTime(json.createdAt),
      updatedAt: GameModel._toDateTime(json.updatedAt),
      createdBy: json.createdBy ?? "",
      category: json.category ?? "",
      tipText: json.tipText ?? "",
      videoDurationSeconds: json.videoDurationSeconds ?? 0,
      levelId: json.levelId ?? "",
    });
  }

  /** Create a copy with modified fields */
  copyWith(updates: Partial<IGameModel>): GameModel {
    return new GameModel({
      ...this,
      ...updates,
    });
  }

  /** Convert Firestore Timestamp or string → Date */
  private static _toDateTime(value: any): Date {
    if (value instanceof Timestamp) return value.toDate();
    if (typeof value === "string") return new Date(value);
    return new Date();
  }
}

File: Models/game models/game_model.dart
import 'package:cloud_firestore/cloud_firestore.dart';

class GameModel {
  final String id;
  final String title;
  final String description;
  final int levelNumber;
  final String videoUrl;
  final String? thumbnailUrl;
  final int estimatedTime;
  final String accessTier;
  final bool isActive;
  final int sortOrder;
  final DateTime createdAt;
  final DateTime updatedAt;
  final String createdBy;
  final String category;
  final String tipText;
  final int videoDurationSeconds;

  const GameModel({
    required this.id,
    required this.title,
    required this.description,
    required this.levelNumber,
    required this.videoUrl,
    this.thumbnailUrl,
    required this.estimatedTime,
    required this.accessTier,
    required this.isActive,
    required this.sortOrder,
    required this.createdAt,
    required this.updatedAt,
    required this.createdBy,
    required this.category,
    required this.tipText,
    required this.videoDurationSeconds,
  });

  factory GameModel.create({
    required String id,
    required String title,
    required String description,
    required int levelNumber,
    required String videoUrl,
    String? thumbnailUrl,
    required int estimatedTime,
    String accessTier = 'free',
    bool isActive = true,
    int sortOrder = 0,
    String createdBy = '',
    required String category,
    required String tipText,
    required int videoDurationSeconds,
  }) {
    final now = DateTime.now();
    return GameModel(
      id: id,
      title: title,
      description: description,
      levelNumber: levelNumber,
      videoUrl: videoUrl,
      thumbnailUrl: thumbnailUrl,
      estimatedTime: estimatedTime,
      accessTier: accessTier,
      isActive: isActive,
      sortOrder: sortOrder,
      createdAt: now,
      updatedAt: now,
      createdBy: createdBy,
      category: category,
      tipText: tipText,
      videoDurationSeconds: videoDurationSeconds,
    );
  }

  Map<String, dynamic> toFirestore() => {
    'id': id,
    'title': title,
    'description': description,
    'levelNumber': levelNumber,
    'videoUrl': videoUrl,
    'thumbnailUrl': thumbnailUrl,
    'estimatedTime': estimatedTime,
    'accessTier': accessTier,
    'isActive': isActive,
    'sortOrder': sortOrder,
    'createdAt': Timestamp.fromDate(createdAt),
    'updatedAt': Timestamp.fromDate(updatedAt),
    'createdBy': createdBy,
    'category': category,
    'tipText': tipText,
    'videoDurationSeconds': videoDurationSeconds,
  };

  factory GameModel.fromFirestore(Map<String, dynamic> json) => GameModel(
    id: json['id'] ?? '',
    title: json['title'] ?? '',
    description: json['description'] ?? '',
    levelNumber: json['levelNumber'] ?? 1,
    videoUrl: json['videoUrl'] ?? '',
    thumbnailUrl: json['thumbnailUrl'],
    estimatedTime: json['estimatedTime'] ?? 0,
    accessTier: json['accessTier'] ?? 'free',
    isActive: json['isActive'] ?? true,
    sortOrder: json['sortOrder'] ?? 0,
    createdAt: _toDateTime(json['createdAt']),
    updatedAt: _toDateTime(json['updatedAt']),
    createdBy: json['createdBy'] ?? '',
    category: json['category'] ?? '',
    tipText: json['tipText'] ?? '',
    videoDurationSeconds: json['videoDurationSeconds'] ?? 0,
  );

  GameModel copyWith({
    String? id,
    String? title,
    String? description,
    int? levelNumber,
    String? videoUrl,
    String? thumbnailUrl,
    int? estimatedTime,
    String? accessTier,
    bool? isActive,
    int? sortOrder,
    DateTime? createdAt,
    DateTime? updatedAt,
    String? createdBy,
    String? category,
    String? tipText,
    int? videoDurationSeconds,
  }) => GameModel(
    id: id ?? this.id,
    title: title ?? this.title,
    description: description ?? this.description,
    levelNumber: levelNumber ?? this.levelNumber,
    videoUrl: videoUrl ?? this.videoUrl,
    thumbnailUrl: thumbnailUrl ?? this.thumbnailUrl,
    estimatedTime: estimatedTime ?? this.estimatedTime,
    accessTier: accessTier ?? this.accessTier,
    isActive: isActive ?? this.isActive,
    sortOrder: sortOrder ?? this.sortOrder,
    createdAt: createdAt ?? this.createdAt,
    updatedAt: updatedAt ?? this.updatedAt,
    createdBy: createdBy ?? this.createdBy,
    category: category ?? this.category,
    tipText: tipText ?? this.tipText,
    videoDurationSeconds: videoDurationSeconds ?? this.videoDurationSeconds,
  );

  static DateTime _toDateTime(dynamic v) {
    if (v is Timestamp) return v.toDate();
    if (v is String) return DateTime.parse(v);
    return DateTime.now();
  }
}
